### ISSUE TITLE
Programmatically trigger buttons?

### ISSUE THREAD

**Issue by Gataquadrada** [nbarkhina/N64Wasm#32]
Hi!

Is there a way to programmatically trigger buttons and sticks?

The only way I managed to trigger commands was by forcing `myApp.mobileMode = true` on Load, and manually calling `mobilePressSelect()` and `mobileReleaseSelect()`.

**Comment by nbarkhina** [comment]
Yes if you are in mobileMode there is a function in `input_controller.js` you can modify to set all the buttons/sticks to whatever you want. You have to leave the function name the same and at the end of the function make sure it calls `sendMobileControls` the way its doing by default, but everything else in the function you can modify to whatever you want. Just make sure you format a `mobileString` with all the buttons in the correct order as 0's and 1's. For the Analog Sticks instead of sending VectorX and VectorY you can send your own values - they need to be a decimal number between -1 and 1 which would be the stick range for both axes.

```
    updateMobileControls(){
        let mobileString = '';
        mobileString += '0'; //UP
        mobileString += '0'; //DOWN
        mobileString += '0'; //LEFT
        mobileString += '0'; //RIGHT
        if (this.Key_Action_A) mobileString += '1'; else mobileString += '0';
        if (this.Key_Action_B) mobileString += '1'; else mobileString += '0';
        if (this.Key_Action_Start) mobileString += '1'; else mobileString += '0';
        if (this.Key_Action_Z && !window["myApp"].rivetsData.useZasCMobile) mobileString += '1'; else mobileString += '0';
        mobileString += '0'; //L
        mobileString += '0'; //R
        if (this.Key_Action_CUP) mobileString += '1'; else mobileString += '0'; //CUP
        if (this.Key_Action_CDOWN) mobileString += '1'; else mobileString += '0'; //CDOWN
        if (this.Key_Action_CLEFT) mobileString += '1'; else mobileString += '0'; //CLEFT
        if (this.Key_Action_CRIGHT) mobileString += '1'; else mobileString += '0'; //CRIGHT

        window["myApp"].sendMobileControls(mobileString, this.VectorX.toString(), this.VectorY.toString());
    }
```

**Comment by Gataquadrada** [comment]
Is there a way to trigger button presses without having to rely on forcing `mobileMode = true`?

**Comment by nbarkhina** [comment]
What you can do is in the function `WriteConfigFile()` in the file `script.js` just modify the part where it's setting the configString based on mobilemode -

So you can change

`if (this.mobileMode) configString += "1" + "\r\n"; else configString += "0" + "\r\n";`

to this

`configString += "1" + "\r\n";;`

That way from the webassembly side it will think's it's in mobileMode but from the javascript/html side it will be in regular mode

**Comment by Gataquadrada** [comment]
Isn't there a simpler way to trigger button presses? 
All of the above seems very convoluted.

My goal is to use a WebSocket to send keyboard key presses.

**Comment by nbarkhina** [comment]
It's just a few lines of code you have to modify, and I'm giving you the exact spots to change. Doesn't seem that convoluted to me. All the plumbing is there already. So whatever comes back from the WebSocket you can just wire up to the `updateMobileControls()` function.

I built the plumbing originally to support Mobile Mode. You're free to change the C++ code to not require mobile mode but I'm trying to save you the trouble of having to re-compile the webassembly part and just stick to the javascript/html side.

**Comment by Gataquadrada** [comment]
I managed to get _most_ buttons to work by changing the boolean value the following group of variables:
`myApp.rivetsData.inputController.Key_Action_{BUTTON_NAME}`

But, for some reason `Key_Action_L` and `Key_Action_R` seem to have no effect.
What would I be missing?

I also noticed the code you pasted above has no check for those buttons, when assembling the `mobileString ` variable.
Is it intentional?
